src/main/rx/rx.c

```diff
@@ -468,18 +470,150 @@ void rxSetUplinkTxPwrMw(uint16_t uplinkTxPwrMwValue)
 }
 #endif
 
+#if defined(USE_PWM) || defined(USE_PPM)
+static uint16_t calculateChannelMovingAverage(uint8_t chan, uint16_t sample)
+{
+    static int16_t rcSamples[MAX_SUPPORTED_RX_PARALLEL_PWM_OR_PPM_CHANNEL_COUNT][PPM_AND_PWM_SAMPLE_COUNT];
+    static int16_t rcDataMean[MAX_SUPPORTED_RX_PARALLEL_PWM_OR_PPM_CHANNEL_COUNT];
+    static bool rxSamplesCollected = false;
+
+    const uint8_t currentSampleIndex = rcSampleIndex % PPM_AND_PWM_SAMPLE_COUNT;
+
+    // update the recent samples and compute the average of them
+    rcSamples[chan][currentSampleIndex] = sample;
+
+    // avoid returning an incorrect average which would otherwise occur before enough samples
+    if (!rxSamplesCollected) {
+        if (rcSampleIndex < PPM_AND_PWM_SAMPLE_COUNT) {
+            return sample;
+        }
+        rxSamplesCollected = true;
+    }
+
+    rcDataMean[chan] = 0;
+    for (int sampleIndex = 0; sampleIndex < PPM_AND_PWM_SAMPLE_COUNT; sampleIndex++) {
+        rcDataMean[chan] += rcSamples[chan][sampleIndex];
+    }
+    return rcDataMean[chan] / PPM_AND_PWM_SAMPLE_COUNT;
+}
+#endif
+
+static uint16_t getRxfailValue(uint8_t channel)
+{
+    const rxFailsafeChannelConfig_t *channelFailsafeConfig = rxFailsafeChannelConfigs(channel);
+
+    switch (channelFailsafeConfig->mode) {
+    case RX_FAILSAFE_MODE_AUTO:
+        switch (channel) {
+        case ROLL:
+        case PITCH:
+        case YAW:
+            return rxConfig()->midrc;
+        case THROTTLE:
+            if (featureIsEnabled(FEATURE_3D) && !IS_RC_MODE_ACTIVE(BOX3D) && !flight3DConfig()->switched_mode3d) {
+                return rxConfig()->midrc;
+            } else {
+                return rxConfig()->rx_min_usec;
+            }
+        }
+
+    FALLTHROUGH;
+    default:
+    case RX_FAILSAFE_MODE_INVALID:
+    case RX_FAILSAFE_MODE_HOLD:
+        return rcData[channel];
+
+    case RX_FAILSAFE_MODE_SET:
+        return RXFAIL_STEP_TO_CHANNEL_VALUE(channelFailsafeConfig->step);
+    }
+}
+
+static void detectAndApplySignalLossBehaviour(void)
```

Comment:
This function ether passes rcRaw information straight through to rcData, or holds/changes the channel values according to failsafe status change.